mdetsims/metacal/metacal_fitter.py

```diff
@@ -0,0 +1,284 @@
+import logging
+import numpy as np
+import esutil as eu
+
+from ngmix.bootstrap import MaxMetacalBootstrapper
+from ngmix.gexceptions import BootPSFFailure, BootGalFailure
+
+from .base_fitter import FitterBase, _fit_one_psf
+from .util import Namer
+
+logger = logging.getLogger(__name__)
+
+METACAL_TYPES = ['noshear', '1p', '1m', '2p', '2m']
+
+
+class MetacalFitter(FitterBase):
+    """Run metacal on all a list of observations.
+
+    Parameters
+    ----------
+    conf : dict
+        A configuration dictionary.
+    nband : int
+        The number of bands.
+    rng : np.random.RandomState
+        An RNG instance.
+    mof_fitter : mdetsims.metcal.MOFFitter
+        A an instantiated MOFFitter for doing neighbor subtraction.
+
+    Methods
+    -------
+    go(mbobs_list_input)
+    """
+    def __init__(self, conf, nband, rng, mof_fitter=None, **kw):
+
+        self.mof_fitter = mof_fitter
+
+        super().__init__(conf, nband, rng)
+
+    def _setup(self):
+        self.metacal_prior = self._get_prior(self['metacal'])
+        assert self.metacal_prior is not None
+        self['metacal']['symmetrize_weight'] = self['metacal'].get(
+            'symmetrize_weight', False)
+
+        # be safe friends!
+        if 'types' in self['metacal']:
+            assert self['metacal']['types'] == METACAL_TYPES
+
+    @property
+    def result(self):
+        """Get the result data"""
+        if not hasattr(self, '_result'):
+            raise RuntimeError('run go() first')
+
+        if self._result is not None:
+            return self._result.copy()
+        else:
+            return None
+
+    def go(self, mbobs_list):
+        """Run metcal on a list of MultiBandObsLists.
+
+        Parameters
+        ----------
+        mbobs_list: list of MultiBandObsList
+            One for each object.  If it is a simple MultiBandObsList it will
+            be converted to a list
+        """
+        if not isinstance(mbobs_list, list):
+            mbobs_list = [mbobs_list]
+
+        if self.mof_fitter is not None:
+            # for mof fitting, we expect a list of mbobs_lists
+            mof_data, epochs_data = self.mof_fitter.go(mbobs_list)
```

Comment:
This block is slightly different since now we store the fitter in a method in the MOF library.

Reply:
OK, as long as it gets run